File: kaliya-server/src/main/java/com/synhaptein/kaliya/modules/reverseIndexer/ReverseIndexer.java
package com.synhaptein.kaliya.modules.reverseIndexer;

import com.synhaptein.kaliya.core.job.Job;
import com.synhaptein.kaliya.core.mapreduce.MapReducer;
import com.synhaptein.kaliya.core.mapreduce.Pair;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 * Demo of a reverse index algorithm.
 *
 * Redistributions of files must retain the above copyright notice.
 *
 * @copyright     Copyright 2010-2011, SynHaptein (http://www.synhaptein.com)
 * @link          http://www.synhaptein.com/kaliya kaliya project
 * @since         kaliya 0.1
 * @license       http://www.synhaptein.com/kaliya/license.html
 */

public class ReverseIndexer extends Job<String, String, List<String>> {
    private List<Pair<String, String>> m_problem;

    public ReverseIndexer(List<Pair<String, String>> p_problem) {
        m_problem = p_problem;
    }

    @Override
    public String resultsToString() {
        String result = "";
        if(getStatus() == JobStatus.FINISHED) {
            for(Pair<String, List<String>> pair : m_results) {
                result += pair.key + ": ";
                for(Iterator<String> it = pair.value.iterator(); it.hasNext();) {
                    result += " " + it.next();
                    if(it.hasNext()) {
                        result += ",";
                    }
                }
                result += "<br/>";
            }
        }
        return result;
    }

    @Override
    public String toString() {
        return "Reverse index of...";
    }

    @Override
    public String getJobName() {
        return "reverseIndexer";
    }

    @Override
    public void initMapReducer(MapReducer<String, String, List<String>> p_mapReducer) {
        //p_mapReducer.setMapOnly();
        //p_mapReducer.setStopOnFirstMap();
    }

    @Override
    public Iterator<Pair<String, String>> getIterator() {
        return m_problem.iterator();
    }
}

File: kaliya-command/src/main/scala/com/synhaptein/kaliya/controller/Reverseindexer.scala
package com.synhaptein.kaliya.controller

import com.synhaptein.scalator.controllers.Controller
import com.synhaptein.kaliya.listener.KaliyaServerListener
import com.synhaptein.scalator.views.{Redirection, View}
import org.codehaus.jackson.`type`.TypeReference
import com.synhaptein.kaliya.core.mapreduce.{Pair, Task}

/**
 * Controller that add a new reverseIndexer job.
 *
 * Redistributions of files must retain the above copyright notice.
 *
 * @copyright     Copyright 2010-2011, SynHaptein (http://www.synhaptein.com)
 * @link          http://www.synhaptein.com/kaliya kaliya project
 * @since         kaliya 0.1
 * @license       http://www.synhaptein.com/kaliya/license.html
 */

class Reverseindexer extends Controller {
  override def index() = {
    val strlstfiles = context.parameter("lstfiles", "")
    try {
      val lstFiles:java.util.List[Pair[String, String]] =
        Task.mapper.readValue(strlstfiles, new TypeReference[java.util.List[Pair[String, String]]] {})
      if(lstFiles.size() == 0) throw new Exception
      KaliyaServerListener.kaliyaServer.getJobScheduler.addJob(
        new com.synhaptein.kaliya.modules.reverseIndexer.ReverseIndexer(lstFiles))
      new View("/console") with Redirection
    }
    catch {
      case _ => new View("/console/addJob?jobtype=reverseIndexer") with Redirection
    }
  }
}